File: 351002/Khmylko/rest-api/publisher/src/Service/TweetService.php
<?php
namespace App\Service;

use App\Exception\ConflictException;
use App\Repository\TweetRepository;
use App\Repository\EditorRepository;
use App\Exception\ApiException;
use App\Exception\ValidationException;
use App\Exception\NotFoundException;

class TweetService {
    private TweetRepository $repository;
    private EditorRepository $editorRepository;
    private MarkerService $markerService;
    public function __construct(TweetRepository $repository, EditorRepository $editorRepo, MarkerService $markerService) {
        $this->repository = $repository;
        $this->editorRepository = $editorRepo;
        $this->markerService = $markerService;
    }
    public function getAll(): array {
        $tweets = $this->repository->findAll();

        // Конвертируем editor_id в editorId для каждого твита
        foreach ($tweets as &$tweet) {
            if (isset($tweet['editor_id'])) {
                $tweet['editorId'] = $tweet['editor_id'];
                unset($tweet['editor_id']);
            }
        }

        return $tweets;
    }
    public function getById(int $id): array {
        $tweet = $this->repository->findById($id);
        if (!$tweet) {
            throw new NotFoundException('Tweet', $id);
        }
        return $tweet;
    }

    public function findAll(int $page = 1, int $limit = 10, string $sortBy = 'id', string $order = 'ASC'): array {
        $tweets = $this->repository->findAll([], $sortBy, $order, $page, $limit);

        // Конвертируем editor_id в editorId для каждого твита
        foreach ($tweets as &$tweet) {
            if (isset($tweet['editor_id'])) {
                $tweet['editorId'] = $tweet['editor_id'];
                unset($tweet['editor_id']);
            }
        }

        return $tweets;
    }

    public function create(array $data): array {
        // Конвертируем editorId в editor_id
        if (isset($data['editorId'])) {
            $data['editor_id'] = $data['editorId'];
        }

        // Валидация
        if (empty($data['title']) || strlen($data['title']) < 2 || strlen($data['title']) > 64) {
            throw new ValidationException("Title must be 2-64 characters");
        }
        if (empty($data['content']) || strlen($data['content']) < 4 || strlen($data['content']) > 2048) {
            throw new ValidationException("Content must be 4-2048 characters");
        }
        if (empty($data['editor_id'])) {
            throw new ValidationException("editorId is required");
        }

        // Проверка существования редактора
        $editor = $this->editorRepository->findById($data['editor_id']);
        if (!$editor) {
            throw new ConflictException("Editor with id {$data['editor_id']} does not exist");
        }

        // Проверка уникальности title для этого редактора
        $existingTweet = $this->repository->findByEditorIdAndTitle($data['editor_id'], $data['title']);
        if ($existingTweet) {
            throw new ConflictException("Tweet with title '{$data['title']}' already exists for this editor");
        }

        // Создаём твит
        $tweet = $this->repository->create($data);
        $tweetId = $tweet['id'] ?? $tweet['tweet_id'] ?? null;

        if (!$tweetId) {
            throw new \Exception("Failed to create tweet");
        }

        // Привязываем маркеры, если они есть
        if (!empty($data['markers']) && is_array($data['markers'])) {
            foreach ($data['markers'] as $markerName) {
                // Ищем или создаём маркер
                $marker = $this->markerService->findOrCreateByName($markerName);

                // Привязываем маркер к твиту
                $this->repository->attachMarker($tweetId, $marker['id']);
            }
        }

        // Конвертируем обратно editor_id в editorId для ответа
        if (isset($tweet['editor_id'])) {
            $tweet['editorId'] = $tweet['editor_id'];
            unset($tweet['editor_id']);
        }

        return $tweet;
    }

    public function update(int $id, array $data): array {
        // Конвертируем editorId в editor_id
        if (isset($data['editorId'])) {
            $data['editor_id'] = $data['editorId'];
        }

        $existing = $this->repository->findById($id);
        if (!$existing) {
            throw new NotFoundException('Tweet', $id);
        }

        if (isset($data['title'])) $existing['title'] = $data['title'];
        if (isset($data['content'])) $existing['content'] = $data['content'];
        if (isset($data['editor_id'])) $existing['editor_id'] = $data['editor_id'];

        $tweet = $this->repository->update($id, $existing);

        // Конвертируем обратно editor_id в editorId для ответа
        if (isset($tweet['editor_id'])) {
            $tweet['editorId'] = $tweet['editor_id'];
            unset($tweet['editor_id']);
        }

        return $tweet;
    }

    public function delete(int $id): void {
        if (!$this->repository->exists($id)) {
            throw new NotFoundException('Tweet', $id);
        }
        $this->repository->delete($id);
    }

    private function validate(array $data): void {
        if (empty($data['title']) || strlen($data['title']) < 2 || strlen($data['title']) > 64) {
            throw new ValidationException("Title must be 2-64 characters");
        }
        if (empty($data['content']) || strlen($data['content']) < 4 || strlen($data['content']) > 2048) {
            throw new ValidationException("Content must be 4-2048 characters");
        }
        if (empty($data['editor_id'])) {
            throw new ValidationException("editorId is required");
        }
    }
}

File: 351003/Mashevskiy/Lab2/src/main/java/by/tracker/rest_api/service/TweetService.java
package by.tracker.rest_api.service;

import by.tracker.rest_api.dto.TweetRequestDto;
import by.tracker.rest_api.dto.TweetResponseDto;
import by.tracker.rest_api.entity.Creator;
import by.tracker.rest_api.entity.Marker;
import by.tracker.rest_api.entity.Tweet;
import by.tracker.rest_api.exception.DuplicateResourceException;
import by.tracker.rest_api.exception.ResourceNotFoundException;
import by.tracker.rest_api.exception.ValidationException;
import by.tracker.rest_api.repository.CreatorRepository;
import by.tracker.rest_api.repository.MarkerRepository;
import by.tracker.rest_api.repository.TweetRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class TweetService {

    @Autowired
    private TweetRepository tweetRepository;

    @Autowired
    private CreatorRepository creatorRepository;

    @Autowired
    private MarkerRepository markerRepository;

    private static final DateTimeFormatter formatter = DateTimeFormatter.ISO_DATE_TIME;

    @Transactional(readOnly = true)
    public TweetResponseDto getById(Long id) {
        Tweet tweet = tweetRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Tweet not found", 40402));
        return toResponseDto(tweet);
    }

    @Transactional(readOnly = true)
    public Page<TweetResponseDto> getAll(Pageable pageable) {
        return tweetRepository.findAll(pageable).map(this::toResponseDto);
    }

    public TweetResponseDto create(TweetRequestDto dto) {
        if (dto.getTitle() == null || dto.getTitle().length() < 2 || dto.getTitle().length() > 64) {
            throw new ValidationException("Title must be between 2 and 64 characters", 40010);
        }
        if (dto.getContent() == null || dto.getContent().length() < 4 || dto.getContent().length() > 2048) {
            throw new ValidationException("Content must be between 4 and 2048 characters", 40011);
        }

        if (tweetRepository.existsByTitle(dto.getTitle())) {
            throw new DuplicateResourceException("Tweet with this title already exists", 40302);
        }

        Creator creator = creatorRepository.findById(dto.getCreatorId())
                .orElseThrow(() -> new ResourceNotFoundException("Creator not found", 40401));

        Tweet tweet = new Tweet();
        tweet.setCreator(creator);
        tweet.setTitle(dto.getTitle());
        tweet.setContent(dto.getContent());

        if (dto.getMarkerIds() != null && !dto.getMarkerIds().isEmpty()) {
            List<Marker> markers = markerRepository.findAllById(dto.getMarkerIds());
            tweet.setMarkers(markers);
        }

        tweet = tweetRepository.save(tweet);
        return toResponseDto(tweet);
    }

    public TweetResponseDto update(TweetRequestDto dto) {
        if (dto.getId() == null) {
            throw new ValidationException("ID is required for update", 40012);
        }

        Tweet existingTweet = tweetRepository.findById(dto.getId())
                .orElseThrow(() -> new ResourceNotFoundException("Tweet not found", 40402));

        if (dto.getTitle() != null && !dto.getTitle().equals(existingTweet.getTitle())) {
            if (tweetRepository.existsByTitle(dto.getTitle())) {
                throw new DuplicateResourceException("Tweet with this title already exists", 40302);
            }
            existingTweet.setTitle(dto.getTitle());
        }

        if (dto.getContent() != null) {
            existingTweet.setContent(dto.getContent());
        }

        if (dto.getCreatorId() != null && !existingTweet.getCreator().getId().equals(dto.getCreatorId())) {
            Creator creator = creatorRepository.findById(dto.getCreatorId())
                    .orElseThrow(() -> new ResourceNotFoundException("Creator not found", 40401));
            existingTweet.setCreator(creator);
        }

        if (dto.getMarkerIds() != null) {
            List<Marker> markers = markerRepository.findAllById(dto.getMarkerIds());
            existingTweet.setMarkers(markers);
        }

        existingTweet = tweetRepository.save(existingTweet);
        return toResponseDto(existingTweet);
    }

    public void delete(Long id) {
        if (!tweetRepository.existsById(id)) {
            throw new ResourceNotFoundException("Tweet not found", 40402);
        }
        tweetRepository.deleteById(id);
    }

    private TweetResponseDto toResponseDto(Tweet entity) {
        TweetResponseDto dto = new TweetResponseDto();
        dto.setId(entity.getId());
        dto.setCreatorId(entity.getCreator() != null ? entity.getCreator().getId() : null);
        dto.setTitle(entity.getTitle());
        dto.setContent(entity.getContent());
        if (entity.getCreated() != null) {
            dto.setCreated(entity.getCreated().format(formatter));
        }
        if (entity.getModified() != null) {
            dto.setModified(entity.getModified().format(formatter));
        }
        if (entity.getMarkers() != null) {
            dto.setMarkerIds(entity.getMarkers().stream().map(Marker::getId).collect(Collectors.toList()));
        }
        return dto;
    }
}